File: packages/core/src/utils/entity-roster.ts
/**
 * 实体名册 + 新专名确认卡（G7b/342 号，Phase B 批次二末项；WNW W6 的采纳）。
 *
 * 防止"同人异名 / 同名异人"：settle 链的章摘要 `characters` 列（LLM 从正文
 * 提取的出场人物）与本报名册逐一比对，未登记项生成确认卡三选——
 * **新实体 / 已有名子的别名 / 疑似笔误**。作者确认前不入册（先例：rename_entity
 * 只有名册内操作；新实体必须过确认卡）。

 * truth 类目 `story/entity_roster.md`（作者可手改，每条一节）：
 *
 *   ## entity-1
 *   - name: 林动
 *   - aliases: 林小哥, 少年林动
 *   - kind: person
 *   - registeredAt: 1
 *
 * 双端：`engine-rs/src/utils/entity_roster.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/entity-roster-vectors.json`。
 */

export interface RosterEntity {
  readonly id: string;
  /** 规范名。 */
  readonly name: string;
  readonly aliases: ReadonlyArray<string>;
  readonly kind: "person" | "place" | "faction" | "item" | "other";
  readonly registeredAt: number;
}

const ENTITY_HEADING = /^##\s+(entity-[\w-]+)\s*$/;
const FIELD_LINE = /^-\s*(\w+)\s*:\s*(.*)$/;
const VALID_KINDS = new Set(["person", "place", "faction", "item", "other"]);
const CHARACTER_SPLIT = /[,，、;；/]/;
const CHARACTER_STOPWORDS = /^(无|没有|暂无|none|n\/a|unknown|未知|others?|其他)$/i;

/** 解析实体名册（手改容错：缺 kind 归 other，坏节跳过）。 */
export function parseEntityRoster(markdown: string): RosterEntity[] {
  const entries: RosterEntity[] = [];
  let current: { id: string; fields: Map<string, string> } | undefined;
  const flush = () => {
    if (!current) return;
    const name = current.fields.get("name")?.trim() ?? "";
    if (!name) {
      current = undefined;
      return;
    }
    const aliases = splitList(current.fields.get("aliases") ?? "");
    const kindRaw = (current.fields.get("kind") ?? "other").trim();
    const kind = (VALID_KINDS.has(kindRaw) ? kindRaw : "other") as RosterEntity["kind"];
    const registeredAt = Number.parseInt((current.fields.get("registeredAt") ?? "").trim(), 10);
    entries.push({
      id: current.id,
      name,
      aliases,
      kind,
      registeredAt: Number.isFinite(registeredAt) ? registeredAt : 0,
    });
    current = undefined;
  };

  for (const rawLine of markdown.split(/\r?\n/)) {
    const line = rawLine.trim();
    const heading = ENTITY_HEADING.exec(line);
    if (heading) {
      flush();
      current = { id: heading[1]!, fields: new Map() };
      continue;
    }
    if (!current) continue;
    const field = FIELD_LINE.exec(line);
    if (field) current.fields.set(field[1]!, field[2] ?? "");
  }
  flush();
  return entries;
}

function splitList(value: string): string[] {
  return value
    .split(/[,，、;；]/)
    .map((item) => item.trim())
    .filter(Boolean);
}

/** 渲染名册（防呆方言：每条一节、平铺 key: value 行）。 */
export function renderEntityRoster(entries: ReadonlyArray<RosterEntity>): string {
  const sections = entries.map((entry) =>
    [
      `## ${entry.id}`,
      `- name: ${entry.name}`,
      `- aliases: ${entry.aliases.join(", ")}`,
      `- kind: ${entry.kind}`,
      `- registeredAt: ${entry.registeredAt}`,
    ].join("\n"),
  );
  return ["# 实体名册（Entity Roster）", ...sections].join("\n\n");
}

/** 从章摘要 characters 列提取人名候选（去重保序、剔除停用词）。 */
export function extractCharacterCandidates(
  summaries: ReadonlyArray<{ readonly chapter: number; readonly characters: string }>,
): string[] {
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const summary of [...summaries].sort((a, b) => a.chapter - b.chapter)) {
    for (const raw of summary.characters.split(CHARACTER_SPLIT)) {
      const name = raw.trim();
      if (!name || CHARACTER_STOPWORDS.test(name)) continue;
      const key = name.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);
      ordered.push(name);
    }
  }
  return ordered;
}

/** 经典 Levenshtein 距离（双端一致的纯 DP 实现）。 */
export function editDistance(a: string, b: string): number {
  const aChars = [...a];
  const bChars = [...b];
  if (aChars.length === 0) return bChars.length;
  if (bChars.length === 0) return aChars.length;
  let previous = Array.from({ length: bChars.length + 1 }, (_, i) => i);
  for (let i = 1; i <= aChars.length; i++) {
    const currentRow = [i];
    for (let j = 1; j <= bChars.length; j++) {
      const substitution = previous[j - 1]! + (aChars[i - 1] === bChars[j - 1] ? 0 : 1);
      currentRow.push(Math.min(previous[j]! + 1, currentRow[j - 1]! + 1, substitution));
    }
    previous = currentRow;
  }
  return previous[bChars.length]!;
}

/** 确认卡三选动作。 */
export type RosterCandidateAction = "new-entity" | "alias" | "typo";

export interface RosterCandidateCard {
  readonly candidate: string;
  readonly action: RosterCandidateAction;
  /** alias/typo 指向的规范名。 */
  readonly targetName?: string;
  readonly confidence: "high" | "medium" | "low";
}

const CONFIDENCE_RANK: Record<RosterCandidateCard["confidence"], number> = {
  high: 0,
  medium: 1,
  low: 2,
};
const ACTION_RANK: Record<RosterCandidateAction, number> = {
  typo: 0,
  alias: 1,
  "new-entity": 2,
};

/**
 * 新专名比对（settle 时调用）：候选与名册逐一比对，未登记项生成确认卡。
 * 判定顺序：
 * 1. 精确命中（name/aliases，大小写不敏感）→ 已登记，不出卡；
 * 2. 与 name 或 alias 编辑距离 ≤1（且候选长度 ≥2）→ typo（指向规范名，high）；
 * 3. 包含关系（候选含 name 或被 name 包含，较短者 ≥2 字）→ alias（指向规范名，medium）；
 * 4. 其余 → new-entity（low）。
 * 排序：action(typo→alias→new) → confidence → candidate 码元序。
 */
export function resolveRosterCandidates(
  candidates: ReadonlyArray<string>,
  roster: ReadonlyArray<RosterEntity>,
): RosterCandidateCard[] {
  const cards: RosterCandidateCard[] = [];
  for (const candidate of candidates) {
    const key = candidate.toLowerCase();
    let matched = false;
    for (const entity of roster) {
      if (entity.name.toLowerCase() === key || entity.aliases.some((a) => a.toLowerCase() === key)) {
        matched = true;
        break;
      }
    }
    if (matched) continue;

    // G7b 判定顺序：包含关系（儿/阿/小缀昵称等）→ 别名；编辑距离 ≤1 → 笔误；
    // 两者都不中 → 新实体。（先昵称后笔误：中文昵称多为后缀增删，会被距离 1 误判）。
    let aliasTarget: string | undefined;
    let typoTarget: string | undefined;
    for (const entity of roster) {
      const nameChars = [...entity.name].length;
      const contained =
        nameChars >= 2 && (candidate.includes(entity.name) || entity.name.includes(candidate));
      if (contained) {
        aliasTarget = entity.name;
        break;
      }
      if (nameChars >= 2 && editDistance(candidate, entity.name) <= 1) {
        typoTarget = entity.name;
        break;
      }
      for (const alias of entity.aliases) {
        if ([...alias].length >= 2 && editDistance(candidate, alias) <= 1) {
          aliasTarget = entity.name;
          break;
        }
      }
      if (aliasTarget) break;
    }
    if (aliasTarget) {
      cards.push({ candidate, action: "alias", targetName: aliasTarget, confidence: "medium" });
      continue;
    }
    if (typoTarget) {
      cards.push({ candidate, action: "typo", targetName: typoTarget, confidence: "high" });
      continue;
    }
    cards.push({ candidate, action: "new-entity", confidence: "low" });
  }
  return cards.sort(
    (a, b) =>
      ACTION_RANK[a.action] - ACTION_RANK[b.action]
      || CONFIDENCE_RANK[a.confidence] - CONFIDENCE_RANK[b.confidence]
      || comparePlain(a.candidate, b.candidate),
  );
}

/** 纯码元比较（禁 localeCompare——排序跨端确定）。 */
function comparePlain(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** 确认动作输入（确认卡三选的落库口径）。 */
export interface RosterConfirmation {
  readonly candidate: string;
  readonly action: RosterCandidateAction;
  readonly targetName?: string;
  /** 登记章（new-entity 用；缺省 0）。 */
  readonly chapter?: number;
}

export interface RosterConfirmationResult {
  readonly roster: ReadonlyArray<RosterEntity>;
  readonly applied:
    | { readonly kind: "new-entity"; readonly id: string }
    | { readonly kind: "alias"; readonly targetName: string }
    | { readonly kind: "typo-ignored" }
    | { readonly kind: "duplicate-ignored" };
}

/**
 * 确认写回（纯函数）：
 * - new-entity → 追加 entity-N+1（registeredAt = chapter ?? 0）；
 * - alias → candidate 并入 targetName 实体的 aliases（去重）；
 * - typo → 不入册（笔误修正属正文修订范畴），kind=typo-ignored；
 * - 重复确认（实体已存在/别名已存在）→ duplicate-ignored。
 */
export function applyRosterConfirmation(
  roster: ReadonlyArray<RosterEntity>,
  confirmation: RosterConfirmation,
): RosterConfirmationResult {
  const key = confirmation.candidate.toLowerCase();
  const alreadyKnown = roster.some(
    (entity) =>
      entity.name.toLowerCase() === key
      || entity.aliases.some((alias) => alias.toLowerCase() === key),
  );
  if (alreadyKnown) {
    return { roster, applied: { kind: "duplicate-ignored" } };
  }

  if (confirmation.action === "new-entity") {
    let maxIndex = 0;
    for (const entity of roster) {
      const match = /^entity-(\d+)$/.exec(entity.id);
      if (match) maxIndex = Math.max(maxIndex, Number.parseInt(match[1]!, 10));
    }
    const entry: RosterEntity = {
      id: `entity-${maxIndex + 1}`,
      name: confirmation.candidate,
      aliases: [],
      kind: "person",
      registeredAt: confirmation.chapter ?? 0,
    };
    return { roster: [...roster, entry], applied: { kind: "new-entity", id: entry.id } };
  }

  if (confirmation.action === "alias" && confirmation.targetName) {
    const target = roster.find((entity) => entity.name === confirmation.targetName);
    if (!target) return { roster, applied: { kind: "duplicate-ignored" } };
    const updated = roster.map((entity) =>
      entity.id === target.id && !entity.aliases.includes(confirmation.candidate)
        ? { ...entity, aliases: [...entity.aliases, confirmation.candidate] }
        : entity,
    );
    return { roster: updated, applied: { kind: "alias", targetName: target.name } };
  }

  return { roster, applied: { kind: "typo-ignored" } };
}

/** 确认卡人话渲染（聊天确认卡/审计注入共用）。 */
export function renderRosterConfirmationCard(
  cards: ReadonlyArray<RosterCandidateCard>,
  language: "zh" | "en" = "zh",
): string | undefined {
  if (cards.length === 0) return undefined;
  const isEn = language === "en";
  const lines = cards.map((card) => {
    if (card.action === "typo") {
      return isEn
        ? `- "${card.candidate}" → possible typo of "${card.targetName}"?`
        : `- "${card.candidate}" → 疑似 "${card.targetName}" 的笔误？`;
    }
    if (card.action === "alias") {
      return isEn
        ? `- "${card.candidate}" → possible alias of "${card.targetName}"?`
        : `- "${card.candidate}" → 可能是 "${card.targetName}" 的别名？`;
    }
    return isEn
      ? `- "${card.candidate}" → new entity?`
      : `- "${card.candidate}" → 新实体？`;
  });
  return [
    isEn ? "Unregistered proper nouns found in this chapter (new entity / alias / typo):" : "本章发现未登记专名（新实体 / 别名 / 笔误）：",
    ...lines,
  ].join("\n");
}

File: engine-rs/src/utils/entity_roster.rs
//! 实体名册 + 新专名确认卡（G7b/342 号，Phase B 批次二末项）。
//!
//! TS 真源：`packages/core/src/utils/entity-roster.ts`；共享向量：
//! `packages/core/src/__tests__/golden/entity-roster-vectors.json`
//! （差分测试 `tests/golden_entity_roster_diff.rs`）。
//!
//! 三选判定顺序：包含关系（昵称后缀）→ alias；编辑距离 ≤1 → typo；其余 new-entity。
//! 精确命中（name/aliases，大小写不敏感）→ 已登记不出卡。

use serde::Deserialize;
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RosterEntity {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    pub kind: String,
    #[serde(default)]
    pub registered_at: i64,
}

const ENTITY_HEADING: &str = "## entity-";

/// 解析实体名册（手改容错：kind 非法归 other，坏节/无名跳过）。
pub fn parse_entity_roster(markdown: &str) -> Vec<RosterEntity> {
    const KINDS: [&str; 5] = ["person", "place", "faction", "item", "other"];
    let mut entries: Vec<RosterEntity> = Vec::new();
    let mut current_id: Option<String> = None;
    let mut fields: std::collections::HashMap<String, String> = std::collections::HashMap::new();
    let mut has_current = false;

    macro_rules! flush {
        () => {
            if has_current {
                let name = fields.get("name").map(|s| s.trim().to_string()).unwrap_or_default();
                if !name.is_empty() {
                    let aliases = split_list(fields.get("aliases").map(String::as_str).unwrap_or(""));
                    let kind_raw = fields.get("kind").map(|s| s.trim().to_string()).unwrap_or_else(|| "other".into());
                    let kind = if KINDS.contains(&kind_raw.as_str()) { kind_raw } else { "other".to_string() };
                    let registered_at = fields
                        .get("registeredAt")
                        .map(String::as_str)
                        .unwrap_or("")
                        .trim()
                        .parse::<i64>()
                        .unwrap_or(0);
                    entries.push(RosterEntity { id: current_id.clone().unwrap_or_default(), name, aliases, kind, registered_at });
                }
            }
            has_current = false;
        };
    }

    for raw_line in markdown.lines() {
        let line = raw_line.trim();
        if let Some(rest) = line.strip_prefix("## ") {
            let name = rest.trim();
            if name.starts_with(ENTITY_HEADING.trim_start_matches("## ")) && name.len() >= 5 {
                flush!();
                current_id = Some(name.to_string());
                has_current = true;
                fields.clear();
                continue;
            }
        }
        if !has_current {
            continue;
        }
        if let Some(rest) = line.strip_prefix("- ") {
            if let Some(colon) = rest.find(':') {
                fields.insert(rest[..colon].trim().to_string(), rest[colon + 1..].trim_start().to_string());
            }
        }
    }
    flush!();
    entries
}

fn split_list(value: &str) -> Vec<String> {
    value
        .split([',', '，', '、', ';', '；'])
        .map(|item| item.trim().to_string())
        .filter(|item| !item.is_empty())
        .collect()
}

/// 渲染名册（防呆方言：每条一节、平铺 key: value 行）。
pub fn render_entity_roster(entries: &[RosterEntity]) -> String {
    let sections: Vec<String> = entries
        .iter()
        .map(|entry| {
            format!(
                "## {}\n- name: {}\n- aliases: {}\n- kind: {}\n- registeredAt: {}",
                entry.id,
                entry.name,
                entry.aliases.join(", "),
                entry.kind,
                entry.registered_at,
            )
        })
        .collect();
    let mut parts = vec!["# 实体名册（Entity Roster）".to_string()];
    parts.extend(sections);
    parts.join("\n\n")
}

/// 从章摘要 characters 列提取人名候选（去重保序、剔除停用词）。
pub fn extract_character_candidates(
    summaries: &[(i64, String)],
) -> Vec<String> {
    const STOPWORDS: [&str; 9] = ["无", "没有", "暂无", "none", "n/a", "unknown", "未知", "other", "others"];
    let mut ordered: Vec<(i64, String)> = summaries.to_vec();
    ordered.sort_by_key(|(chapter, _)| *chapter);
    let mut seen: Vec<String> = Vec::new();
    let mut out: Vec<String> = Vec::new();
    for (_, characters) in ordered {
        for raw in characters.split([',', '，', '、', ';', '；', '/']) {
            let name = raw.trim();
            if name.is_empty() || STOPWORDS.iter().any(|s| name.eq_ignore_ascii_case(s)) {
                continue;
            }
            let key = name.to_lowercase();
            if seen.contains(&key) {
                continue;
            }
            seen.push(key);
            out.push(name.to_string());
        }
    }
    out
}

/// 经典 Levenshtein 距离（双端一致的纯 DP 实现）。
pub fn edit_distance(a: &str, b: &str) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    if a.is_empty() {
        return b.len();
    }
    if b.is_empty() {
        return a.len();
    }
    let mut previous: Vec<usize> = (0..=b.len()).collect();
    for i in 1..=a.len() {
        let mut current_row = vec![i];
        for j in 1..=b.len() {
            let substitution = previous[j - 1] + usize::from(a[i - 1] != b[j - 1]);
            current_row.push((previous[j] + 1).min(current_row[j - 1] + 1).min(substitution));
        }
        previous = current_row;
    }
    previous[b.len()]
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RosterCandidateCard {
    pub candidate: String,
    pub action: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_name: Option<String>,
    pub confidence: &'static str,
}

fn action_rank(action: &str) -> u8 {
    match action {
        "typo" => 0,
        "alias" => 1,
        _ => 2,
    }
}

fn confidence_rank(confidence: &str) -> u8 {
    match confidence {
        "high" => 0,
        "medium" => 1,
        _ => 2,
    }
}

/// 新专名比对（settle 时调用）：判定顺序 包含关系→alias / 编辑距离≤1→typo / 其余 new。
pub fn resolve_roster_candidates(
    candidates: &[String],
    roster: &[RosterEntity],
) -> Vec<RosterCandidateCard> {
    let mut cards: Vec<RosterCandidateCard> = Vec::new();
    for candidate in candidates {
        let key = candidate.to_lowercase();
        let matched = roster.iter().any(|entity| {
            entity.name.to_lowercase() == key
                || entity.aliases.iter().any(|alias| alias.to_lowercase() == key)
        });
        if matched {
            continue;
        }

        let mut alias_target: Option<String> = None;
        let mut typo_target: Option<String> = None;
        for entity in roster {
            let name_chars = entity.name.chars().count();
            if name_chars >= 2
                && (candidate.contains(&entity.name) || entity.name.contains(candidate.as_str()))
            {
                alias_target = Some(entity.name.clone());
                break;
            }
            if name_chars >= 2 && edit_distance(candidate, &entity.name) <= 1 {
                typo_target = Some(entity.name.clone());
                break;
            }
            for alias in &entity.aliases {
                if alias.chars().count() >= 2 && edit_distance(candidate, alias) <= 1 {
                    alias_target = Some(entity.name.clone());
                    break;
                }
            }
            if alias_target.is_some() {
                break;
            }
        }
        if let Some(target) = alias_target {
            cards.push(RosterCandidateCard {
                candidate: candidate.clone(),
                action: "alias",
                target_name: Some(target),
                confidence: "medium",
            });
            continue;
        }
        if let Some(target) = typo_target {
            cards.push(RosterCandidateCard {
                candidate: candidate.clone(),
                action: "typo",
                target_name: Some(target),
                confidence: "high",
            });
            continue;
        }
        cards.push(RosterCandidateCard {
            candidate: candidate.clone(),
            action: "new-entity",
            target_name: None,
            confidence: "low",
        });
    }
    cards.sort_by(|a, b| {
        action_rank(a.action)
            .cmp(&action_rank(b.action))
            .then(confidence_rank(a.confidence).cmp(&confidence_rank(b.confidence)))
            .then(a.candidate.cmp(&b.candidate))
    });
    cards
}

/// 确认卡人话渲染（聊天确认卡/审计注入共用）；空卡返回 None。
pub fn render_roster_confirmation_card(
    cards: &[RosterCandidateCard],
    language: &str,
) -> Option<String> {
    if cards.is_empty() {
        return None;
    }
    let is_en = language == "en";
    let mut lines = vec![
        if is_en {
            "Unregistered proper nouns found in this chapter (new entity / alias / typo):".to_string()
        } else {
            "本章发现未登记专名（新实体 / 别名 / 笔误）：".to_string()
        },
    ];
    for card in cards {
        let line = if card.action == "typo" {
            if is_en {
                format!("- \"{}\" → possible typo of \"{}\"?", card.candidate, card.target_name.clone().unwrap_or_default())
            } else {
                format!("- \"{}\" → 疑似 \"{}\" 的笔误？", card.candidate, card.target_name.clone().unwrap_or_default())
            }
        } else if card.action == "alias" {
            if is_en {
                format!("- \"{}\" → possible alias of \"{}\"?", card.candidate, card.target_name.clone().unwrap_or_default())
            } else {
                format!("- \"{}\" → 可能是 \"{}\" 的别名？", card.candidate, card.target_name.clone().unwrap_or_default())
            }
        } else if is_en {
            format!("- \"{}\" → new entity?", card.candidate)
        } else {
            format!("- \"{}\" → 新实体？", card.candidate)
        };
        lines.push(line);
    }
    Some(lines.join("\n"))
}

/// 机器可读契约（双端 golden 锁形状）。
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityRosterContract {
    pub actions: Vec<&'static str>,
    pub kinds: Vec<&'static str>,
    pub truth_file: &'static str,
    #[serde(rename = "typoMaxDistance")]
    pub typo_max_distance: usize,
    #[serde(rename = "containmentMinChars")]
    pub containment_min_chars: usize,
}

pub fn entity_roster_contract() -> EntityRosterContract {
    EntityRosterContract {
        actions: vec!["new-entity", "alias", "typo"],
        kinds: vec!["person", "place", "faction", "item", "other"],
        truth_file: "story/entity_roster.md",
        typo_max_distance: 1,
        containment_min_chars: 2,
    }
}
